04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        "\n",
+        "За да постигнем тази цел, ни трябват две неща - функция, която да \"спре\" изпълнението до дадено място и да ни върне някаква стойност - функцията, която извършва това, е `yield`. "
+      ],
+      "metadata": {
+        "id": "6jPtmYzPGz4Z"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def perfect_squares(n):\n",
+        "    for i in range(1, n+1):\n",
+        "        yield generate_nth_perfect_square(i)\n",
+        "\n",
+        "\n",
+        "print(perfect_squares(5))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "0coNKoouH2GS",
+        "outputId": "47a76f12-74b3-4da8-b8d5-0903e33c5603"
+      },
+      "execution_count": 26,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "<generator object perfect_squares at 0x7f2c9d0b8350>\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Тук виждаме, че нашата функция вече ни връща `generator` обект. \n",
+        "\n",
+        "За да работим с нашия генератор обект, е необходимо да го подадем като обект на функцията `next` - тя ни връща следващия елемент от генератор/итератор."
+      ],
+      "metadata": {
+        "id": "J8UJgR5QIFI0"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "perfect_square_generator = perfect_squares(5)\n",
+        "print(next(perfect_square_generator))\n",
+        "print(next(perfect_square_generator))\n",
+        "print(next(perfect_square_generator))\n",
+        "print(next(perfect_square_generator))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "7-jHEPqyIwxB",
+        "outputId": "e51fc5f3-1422-405b-d519-dc8d98879cb5"
+      },
+      "execution_count": 27,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "0\n",
+            "1\n",
+            "4\n",
+            "9\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Както се вижда, `next` изпълнява нашия код, до следващото срещане на `yield`, и връща стойността, която `yield` връща."
+      ],
+      "metadata": {
+        "id": "mscLIoUaJOhH"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Благодарение на факта, че получваме генератор, може да използваме нашата функция `perfect_squares` директно във `for` цикъл. Начина по който `for` цикъла работи е, че извиква `next` метода на подадения от нас обект. "
```

Comment:
I would start by emphasizing why we would need iterators, i.e. infinite loops, and then show that they are interchangeable with for loops. Kind of following the approach:
1. We got a problem
2. How to solve it
3. Ok that new thing we used is actually quite similar to something we already know, but it's more powerful

Reply:
Don't know if this is the place to dive into the semantics of iteration. Really don't want to go deep here.